File: ShareOCDemo/ShareOCDemo/UIWebViewVC.m
//
//  UIWebViewVC.m
//  ShareOCDemo
//
//  Created by Miaoz on 2018/6/26.
//  Copyright © 2018年 Miaoz. All rights reserved.
//

#import "UIWebViewVC.h"
#import <JavaScriptCore/JavaScriptCore.h>
#import <UMCommon/UMCommon.h>
#import <UMShare/UMShare.h>
#import <UShareUI/UShareUI.h>

@interface UIWebViewVC ()<UIWebViewDelegate>
@property (nonatomic, strong) JSContext *jsContext;
@end

@implementation UIWebViewVC

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view.
    self.title = @"UIWebViewVC";
    NSURL *url = nil;
    if ([self.startUrlStr isEqualToString:@"www/index.html"]) {
        url = [NSURL fileURLWithPath: [[NSBundle mainBundle] pathForResource:@"index.html" ofType:nil]];
    }
    else{
        url = [NSURL URLWithString:self.startUrlStr && self.startUrlStr.length > 0 ? self.startUrlStr: nil];
    }
    
    CGRect bouds = [[UIScreen mainScreen] applicationFrame];
    UIWebView* webView = [[UIWebView alloc]initWithFrame:bouds];
    webView.delegate = self;
    [self.view addSubview:webView];
    NSURLRequest* request = [NSURLRequest requestWithURL:url];//创建
    [webView loadRequest:request];//加载

}


#pragma mark --UIWebViewDelegate

- (BOOL)webView:(UIWebView *)webView shouldStartLoadWithRequest:(NSURLRequest *)request navigationType:(UIWebViewNavigationType)navigationType {
    
    return YES;
}
- (void)webViewDidStartLoad:(UIWebView *)webView {
    
}
- (void)webViewDidFinishLoad:(UIWebView *)webView {

    self.jsContext = (JSContext *)[webView valueForKeyPath:@"documentView.webView.mainFrame.javaScriptContext"];
    self.jsContext.exceptionHandler = ^(JSContext *con, JSValue *exception) {
        NSLog(@"%@", exception);
        con.exception = exception;
    };
   
    __weak __typeof(self) weakSelf = self;
    // @"executeH5InfoAction"为和js 约定执行的方法名,这样js 可以自由选择时机调用native端 OC 代码
      self.jsContext[@"executeH5InfoAction"] = ^(id info) {
          __strong __typeof(self) strongSelf = weakSelf;
          //回调js 传递的参数 和上边info一样
//        NSArray *args = [JSContext currentArguments];
          dispatch_async(dispatch_get_main_queue(), ^{
                [strongSelf getDataInfo:(NSDictionary *)info];
        });
    };
    

}
- (void)webView:(UIWebView *)webView didFailLoadWithError:(NSError *)error{
    
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

/*
#pragma mark - Navigation

// In a storyboard-based application, you will often want to do a little preparation before navigation
- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
    // Get the new view controller using [segue destinationViewController].
    // Pass the selected object to the new view controller.
}
*/

@end

File: ShareSwiftDemo/ShareSwiftDemo/UIWebViewVC.swift
//
//  UIWebViewVC.swift
//  ShareSwiftDemo
//
//  Created by Miaoz on 2018/6/26.
//  Copyright © 2018年 Miaoz. All rights reserved.
//

import Foundation
import JavaScriptCore

class UIWebViewVC: BaseViewController {
    var startUrlStr:String?
    var jsContext:JSContext?
    override func viewDidLoad() {
        super.viewDidLoad()
        // Do any additional setup after loading the view, typically from a nib.
        var url:URL?
        if self.startUrlStr == "www/index.html" {
            url = URL.init(fileURLWithPath: Bundle.main.path(forResource: "index.html", ofType: nil)!)
//            fileURL(withPath:Bundle.main.path(forResource: "index.html", ofType: nil)!) as NSURL
        } else {
            url = URL(string: ((self.startUrlStr != nil) && (self.startUrlStr?.count)! > 0 ? self.startUrlStr:nil)!)
        }
        
        let webView = UIWebView(frame:  UIScreen.main.applicationFrame)
        webView.delegate = (self as! UIWebViewDelegate)
        self.view.addSubview(webView)

        let request = NSURLRequest(url: url! as URL)
        webView.loadRequest(request as URLRequest)
        
    }
    
   
    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }
}


@objc protocol JavaScriptSwiftDelegate: JSExport {
    
    func executeH5InfoAction(_ info:NSDictionary)
    //注意：参数前面需要加 "_" 下划线
}
@objc class SwiftJavaScriptModel: NSObject, JavaScriptSwiftDelegate {
    
    weak var controller: BaseViewController?
    weak var jsContext: JSContext?
    
    func executeH5InfoAction(_ info:NSDictionary) {
        DispatchQueue.main.async(execute: {
                self.controller!.getDataInfo(info: info)
            })
    }
}

extension UIWebViewVC:UIWebViewDelegate {
    
   
    public func webView(_ webView: UIWebView, shouldStartLoadWith request: URLRequest, navigationType: UIWebViewNavigationType) -> Bool {
        
        return true
    }
    
    public func webViewDidStartLoad(_ webView: UIWebView) {
        
    }
    
    public func webViewDidFinishLoad(_ webView: UIWebView) {
        self.jsContext = webView.value(forKeyPath: "documentView.webView.mainFrame.javaScriptContext") as? JSContext
        self.jsContext?.exceptionHandler = { con, exception in
            if let anException = exception {
                print("\(anException)")
            }
            if let anException = exception {
                con?.exception = anException
            }
        }
        
        let model = SwiftJavaScriptModel()
        model.controller = self
        model.jsContext = self.jsContext
        self.jsContext?.setObject(model, forKeyedSubscript: "Native" as NSCopying & NSObjectProtocol)
        //swift不能使用这种，只能使用model桥接
        // @"executeH5InfoAction"为和js 约定执行的方法名,这样js 可以自由选择时机调用native端 OC 代码
//        self.jsContext!["executeH5InfoAction"] = { [weak self] info in
//            DispatchQueue.main.async(execute: {
//                self.getDataInfo(info as? [AnyHashable: Any])
//            })
//        }


    }
    
    public func webView(_ webView: UIWebView, didFailLoadWithError error: Error) {
        
    }
}

